### ISSUE TITLE
Arbitrary folder structure for Config Context retrieved from Git repository

### ISSUE THREAD

**Issue by chadell** [nautobot/nautobot#1189]
### Environment
* Python version: 3.6.2
* Nautobot version: 1.2.1

### Proposed Functionality

When importing Config Context data from Git repositories, instead of assuming flat folder structure within the [key folder names](https://github.com/nautobot/nautobot/blob/develop/nautobot/extras/datasources/git.py#L294), we leave it up to the user to define any arbitrary folder structure, and via recursive function we extract all the files.

### Use Case

When retrieving Config Context from a Git repository, we look for some [key folder names](https://github.com/nautobot/nautobot/blob/develop/nautobot/extras/datasources/git.py#L270) (such as `devices`, `regions`, etc.) and then within them, we look for specific matches of the `slug` in YAML or JSON files.
When you manage thousands of elements, the list gets big and updating data could be misleading, so being able to create any kind of folder structure within these folders could help to a more user-friendly composition.
For instance, the user could decide to group all the devices by the region and the site it belongs too, for easier management:
Today we have:
```
├── devices
│   ├── device_1.yaml
│   ├── device_2.yaml
│   ├── device_3.yaml
│   ├── device_4.yaml
```

With this feature, the user could define their own folder structure:
```
├── devices
│   ├── abc
│   │   └── device_2.yaml
│   │   └── 123
│   │   │   ├── device_3.yaml
│   ├── xyz
│   │   └── device_1.yaml
│   │   ├── device_4.yaml
```

### Database Changes

No changes expected

### External Dependencies

No changes expected

**Comment by bryanculver** [comment]
@chadell Validating a constraint: The folders `abc`, `123`, and `xyz` in your example above should not follow any convention and are therefor arbitrary. The system would just search all subdirectories under `devices` for all device configs.

**Comment by chadell** [comment]
yes @bryanculver , the folders are totally arbitrary, the more convenient for the user purposes. It's kind of flatting all the files within `devices`, still using the filename for mapping.

**Comment by chadell** [comment]
something like [this](https://github.com/nautobot/nautobot-plugin-golden-config/blob/develop/nautobot_golden_config/datasources.py#L157-L162) could do the trick
